compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/DefaultArgumentStubGenerator.kt

```diff
@@ -415,181 +308,78 @@ class DefaultParameterCleaner constructor(val context: CommonBackendContext) : F
-    }
-
-    if (this is IrConstructor) {
-        syntheticParameters += newFunction.valueParameter(
-            syntheticParameters.last().index + 1,
-            kConstructorMarkerName,
-            context.ir.symbols.defaultConstructorMarker.owner.defaultType.makeNullable()
-        )
-    } else if (context.ir.shouldGenerateHandlerParameterForDefaultBodyFun()) {
-        syntheticParameters += newFunction.valueParameter(
-            syntheticParameters.last().index + 1,
-            "handler".synthesizedName,
-            context.irBuiltIns.anyNType
-        )
-    }
-
-    newFunction.copyTypeParametersFrom(this)
-    val newValueParameters = valueParameters.map { it.copyMaybeNullableTo(newFunction, context) } + syntheticParameters
-    newValueParameters.forEach {
-        it.defaultValue = null
-    }
-
-    newFunction.returnType = returnType
-    newFunction.dispatchReceiverParameter = dispatchReceiverParameter?.copyTo(newFunction)
-    newFunction.extensionReceiverParameter = extensionReceiverParameter?.copyTo(newFunction)
-    newFunction.valueParameters += newValueParameters
-
-    annotations.mapTo(newFunction.annotations) { it.deepCopyWithSymbols() }
-
-    if (origin == IrDeclarationOrigin.FAKE_OVERRIDE) {
-        for (baseFunSymbol in (this as IrSimpleFunction).overriddenSymbols) {
-            val baseFun = baseFunSymbol.owner
-            if (baseFun.needsDefaultArgumentsLowering(skipInlineMethods, skipExternalMethods)) {
-                val baseOrigin = if (baseFun.valueParameters.any { it.defaultValue != null }) {
-                    IrDeclarationOrigin.FUNCTION_FOR_DEFAULT_PARAMETER
-                } else {
-                    IrDeclarationOrigin.FAKE_OVERRIDE
-                }
-                val defaultsBaseFun = baseFun.generateDefaultsFunction(context, baseOrigin, skipInlineMethods, skipExternalMethods)
-                (newFunction as IrSimpleFunction).overriddenSymbols.add((defaultsBaseFun as IrSimpleFunction).symbol)
-            }
+): IrFunction? = context.ir.defaultParameterDeclarationsCache[this] ?: when {
+    skipInlineMethods && isInline -> null
+    skipExternalMethods && isExternalOrInheritedFromExternal() -> null
+    valueParameters.any { it.defaultValue != null } ->
+        generateDefaultsFunctionImpl(context, IrDeclarationOrigin.FUNCTION_FOR_DEFAULT_PARAMETER).also {
+            context.ir.defaultParameterDeclarationsCache[this] = it
+        }
+    this is IrSimpleFunction -> {
+        // If this is an override of a function with default arguments, produce a fake override of a default stub.
+        val overriddenStubs = overriddenSymbols.mapNotNull {
+            it.owner.generateDefaultsFunction(context, skipInlineMethods, skipExternalMethods)?.symbol as IrSimpleFunctionSymbol?
         }
+        if (overriddenStubs.isNotEmpty())
+            generateDefaultsFunctionImpl(context, IrDeclarationOrigin.FAKE_OVERRIDE).also {
```

Comment:
I just realized that both the original version and the refactored one would generate fake overrides with all arguments non-nullable, as only the implementation had `defaultValue`s on the original function's arguments. This does not actually affect anything because `parametersForCall` resolves fake overrides, but it may be something to keep in mind.

Reply:
Thanks, indeed, it makes sense to remember this.